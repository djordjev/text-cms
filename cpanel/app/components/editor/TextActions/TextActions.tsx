import { IconLink } from '@tabler/icons-react';
import classnames from 'classnames';
import { FC, useState } from 'react';
import { Editor } from 'slate';
import { useSlate } from 'slate-react';

import { TextAction } from '~/components/modals/TextAction';
import { Action, ClickActionPayload } from '~/types/editor';

export interface TextActionsProps {
  className?: string;
  onClick: (payload?: ClickActionPayload) => void;
}

const TextActions: FC<TextActionsProps> = (props) => {
  const { className, onClick } = props;

  // Hooks
  const editor = useSlate();
  const [open, setOpen] = useState(false);

  // Setup
  const marks = Editor.marks(editor);
  const active = !!marks?.[Action.Click];

  // Styles
  const classes = classnames('u-p-1xs', className, {
    'u-bg-accent u-rounded-md': active
  });

  // Handlers
  const onToggle = () => {
    if (active) {
      onClick();
      return;
    }

    setOpen(true);
  };

  const onComplete = (payload?: ClickActionPayload) => {
    setOpen(false);

    if (!payload) return;

    onClick(payload);
  };

  // Markdown

  // Life-cycle

  // Short-circuit

  return (
    <>
      <button className={classes} onClick={onToggle} type="button">
        <IconLink />
      </button>

      <TextAction open={open} onComplete={onComplete} />
    </>
  );
};

export { TextActions };
