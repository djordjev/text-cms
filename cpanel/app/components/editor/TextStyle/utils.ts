import { BaseEditor, Editor, Element } from 'slate';
import { ReactEditor } from 'slate-react';

import { CustomElement } from '~/components/editor/VariationEditor/types';

const getSelectedStyle = (editor: BaseEditor & ReactEditor) => {
  const { selection } = editor;

  if (!selection) return;

  const [match] = Array.from(
    Editor.nodes<CustomElement>(editor, {
      at: selection,
      match: (n) =>
        !Editor.isEditor(n) &&
        Element.isElement(n) &&
        (n.type === 'paragraph' || n.type === 'heading')
    })
  );

  const [nodeType] = match;

  return nodeType;
};

export { getSelectedStyle };
