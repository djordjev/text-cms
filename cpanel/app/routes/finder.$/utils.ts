import { CreateParam } from '~/types';

const getCreateOpenModal = (params: URLSearchParams): CreateParam | null => {
  const param = params.get('action');

  if (param === 'new-file' || param === 'new-folder') return param;

  return null;
};

export { getCreateOpenModal };
