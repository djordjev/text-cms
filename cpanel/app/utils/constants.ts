export const COLORS = {
  BLACK: '#000',
  PRIMARY: '#003EB3'
};

export const BUTTON_ACTION = '_action';
