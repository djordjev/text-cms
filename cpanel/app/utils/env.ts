const readEnvVariable = <T = string>(name: string) => {
  const vars = process.env;

  const val = vars[name];
  if (!val) throw new Error(`Variable ${name} not supplied`);

  return val as T;
};

export { readEnvVariable };
