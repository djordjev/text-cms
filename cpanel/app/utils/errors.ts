const buildErrorResponse = (status = 404, message = 'Not found') => {
  return new Response(message, { status, statusText: message });
};

export { buildErrorResponse };
