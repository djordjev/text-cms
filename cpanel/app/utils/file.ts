const isHome = (name: string) => {
  return name.toLowerCase() === 'finder';
};

const isFile = (name: string) => {
  return name.endsWith('.txt');
};

const isFolder = (name: string) => {
  return !isFile(name);
};

const getPathFromSegments = (segments: string[], segment: string) => {
  const index = segments.indexOf(segment);

  if (index < 1) return '/finder';

  const mapped = segments.slice(0, index + 1).map((segment) => {
    if (isHome(segment)) return 'finder';

    return segment;
  });

  const joined = mapped.join('/');

  return `/${joined}`;
};

export { getPathFromSegments, isHome, isFile, isFolder };
