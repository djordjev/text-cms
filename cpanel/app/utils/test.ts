import { Descendant } from 'slate';

export const TEST_EDITOR: Descendant[] = [
  {
    type: 'heading',
    children: [
      {
        text: 'HEADING'
      }
    ],
    level: 3,
    align: 'center'
  },
  {
    type: 'paragraph',
    children: [
      {
        text: 'A '
      },
      {
        text: 'line',
        italic: true
      },
      {
        text: ' of '
      },
      {
        text: 'text',
        color: '#e41111',
        bold: true
      },
      {
        text: ' in a '
      },
      {
        text: 'paragraph',
        bold: true
      },
      {
        text: '.'
      }
    ]
  }
];

export const buildRequest = (data: Record<string, string>, method = 'POST') => {
  const form = new URLSearchParams(data);

  return new Request('https://www.test.local', { body: form, method });
};
