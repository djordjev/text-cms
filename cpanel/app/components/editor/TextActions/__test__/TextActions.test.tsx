const mocks = vi.hoisted(() => ({ TextAction: (props: {open: boolean}) => props.open && <div data-testid="modal" /> })); // prettier-ignore
vi.mock('~/components/modals/TextAction', () => ({ ...mocks }));

import { createRemixStub } from '@remix-run/testing';
import { fireEvent, render, screen } from '@testing-library/react';
import { createEditor } from 'slate';
import { Slate, withReact } from 'slate-react';

import { TEST_EDITOR } from '~/utils/test';

import { TextActions, TextActionsProps } from '../TextActions';

test('TextActions', () => {
  const onClick = vi.fn();

  const props: TextActionsProps = { onClick };

  const editor = withReact(createEditor());

  const createComponent = () => {
    const RemixStub = createRemixStub([
      {
        path: '/',
        Component: () => (
          <Slate editor={editor} initialValue={TEST_EDITOR}>
            <TextActions {...props} />
          </Slate>
        )
      }
    ]);

    return <RemixStub />;
  };

  const { container } = render(createComponent());

  expect(container.childNodes).toMatchSnapshot('initial');

  fireEvent.click(screen.getByRole('button'));

  expect(screen.getByTestId('modal')).toBeInTheDocument();
});
