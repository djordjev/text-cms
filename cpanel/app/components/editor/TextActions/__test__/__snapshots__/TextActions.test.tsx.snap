// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextActions > initial 1`] = `
NodeList [
  <button
    class="u-p-1xs"
    type="button"
  >
    <svg
      class="tabler-icon tabler-icon-link"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9 15l6 -6"
      />
      <path
        d="M11 6l.463 -.536a5 5 0 0 1 7.071 7.072l-.534 .464"
      />
      <path
        d="M13 18l-.397 .534a5.068 5.068 0 0 1 -7.127 0a4.972 4.972 0 0 1 0 -7.071l.524 -.463"
      />
    </svg>
  </button>,
]
`;
